### PR TITLE
Samza on Kafka 0.10

### PR DESCRIPTION
**Pull request by ncolomer**
This PR aims to make Samza run with Kafka 0.10.

Related material:
- [SAMZA-855](https://issues.apache.org/jira/browse/SAMZA-855) - Upgrade Samza's Kafka client version to 0.10.0.0
- [RB-49212](https://reviews.apache.org/r/49212) - RFC: SAMZA-855: Update kafka client to 0.10.0.0

Orignal [patch](https://reviews.apache.org/r/49212/diff/1#4) submitted by Robert Crim was used and rebased on current Samza's master branch.

Next steps:
- [x] make unit tests pass using 
  
  ``` bash
  ./gradlew clean test
  ```
- [x] make integration tests pass using 
  See [Running Integration Tests Locally](http://samza.apache.org/contribute/tests.html) and [Setup passphraseless ssh to localhost on OS X](http://stackoverflow.com/a/7134542/739479)
  
  ``` bash
  # Enable no-password self-login ssh
  ssh-copy-id -i ~/.ssh/id_rsa.pub localhost
  # Check JAVA_HOME is correctly set (if not and you run on zsh, edit ~/.zshenv and add it)
  ssh localhost 'echo $JAVA_HOME'
  # Run integration tests
  ./bin/integration-tests.sh /tmp/samza-tests-$(date +%Y%m%d%H%M%S) --nopassword
  # In case of failure, make sure all sytem ports are available on your hosts
  ```
  
  image
  
  Here are tests logs.
  
  ```
  2016-09-23 14:43:26 zopkio [INFO] Starting zopkio
  2016-09-23 14:43:26 naarad [INFO] Available graphing modules:  
  2016-09-23 14:43:26 zopkio.test_runner [INFO] Setting up configuration: single execution
  2016-09-23 14:43:26 deployment [INFO] Current working directory: /private/tmp/samza-tests-20160923144307
  2016-09-23 14:43:26 deployment [INFO] Downloading: https://archive.apache.org/dist/hadoop/common/hadoop-2.6.1/hadoop-2.6.1.tar.gz
  2016-09-23 14:44:08 deployment [INFO] Downloading: http://archive.apache.org/dist/kafka/0.10.0.1/kafka_2.10-0.10.0.1.tgz
  2016-09-23 14:44:15 deployment [INFO] Downloading: http://archive.apache.org/dist/zookeeper/zookeeper-3.4.6/zookeeper-3.4.6.tar.gz
  2016-09-23 14:44:20 deployment [INFO] Deploying zookeeper_instance_0 on host: localhost
  2016-09-23 14:44:22 deployment [INFO] Deploying yarn_rm_instance_0 on host: localhost
  2016-09-23 14:44:27 deployment [INFO] Deploying yarn_nm_instance_0 on host: localhost
  2016-09-23 14:44:33 deployment [INFO] Deploying kafka_instance_0 on host: localhost
  2016-09-23 14:44:34 samza_job_yarn_deployer [INFO] Deploying tests on host: localhost
  2016-09-23 14:44:37 samza_job_yarn_deployer [INFO] Installing yarn-site.xml to /private/tmp/samza-tests-20160923144307/tests/config/yarn-site.xml
  2016-09-23 14:44:37 smoke_tests [INFO] Running test_samza_job
  2016-09-23 14:44:37 kafka.client [ERROR] Error loading topic metadata for samza-test-topic: <class 'kafka.errors.LeaderNotAvailableError'> (5)
  2016-09-23 14:44:38 util [INFO] Starting tests.negate_number
  2016-09-23 14:44:41 util [INFO] Awaiting tests.negate_number
  2016-09-23 14:44:58 util [INFO] Starting tests.container-performance
  2016-09-23 14:45:02 util [INFO] Awaiting tests.container-performance
  2016-09-23 14:46:13 performance_tests [INFO] Running test_kafka_read_write_performance
  2016-09-23 14:46:13 kafka.client [ERROR] Error loading topic metadata for kafka-read-write-performance-input: <class 'kafka.errors.LeaderNotAvailableError'> (5)
  2016-09-23 14:46:14 performance_tests [INFO] Loading 1000000 test messages.
  2016-09-23 14:46:14 performance_tests [INFO] Loaded 0 messages.
  2016-09-23 14:46:42 performance_tests [INFO] Loaded 100000 messages.
  2016-09-23 14:47:10 performance_tests [INFO] Loaded 200000 messages.
  2016-09-23 14:47:37 performance_tests [INFO] Loaded 300000 messages.
  2016-09-23 14:48:04 performance_tests [INFO] Loaded 400000 messages.
  2016-09-23 14:48:30 performance_tests [INFO] Loaded 500000 messages.
  2016-09-23 14:48:57 performance_tests [INFO] Loaded 600000 messages.
  2016-09-23 14:49:24 performance_tests [INFO] Loaded 700000 messages.
  2016-09-23 14:49:53 performance_tests [INFO] Loaded 800000 messages.
  2016-09-23 14:50:21 performance_tests [INFO] Loaded 900000 messages.
  2016-09-23 14:50:46 util [INFO] Starting tests.kafka-read-write-performance
  2016-09-23 14:50:50 util [INFO] Awaiting tests.kafka-read-write-performance
  2016-09-23 14:51:33 zopkio.deployer [ERROR] Log file /tmp/samza-tests-20160923144307/deploy/kafka/log-cleaner.log does not exist on localhost
  2016-09-23 14:51:33 naarad.utils [WARNING] /tmp/samza-tests-20160923144307/scripts/naarad.cfg : file does not exist.
  2016-09-23 14:51:33 naarad.utils [WARNING] /tmp/samza-tests-20160923144307/scripts/naarad.cfg : file does not exist.
  2016-09-23 14:51:33 naarad.utils [WARNING] /tmp/samza-tests-20160923144307/scripts/naarad.cfg : file does not exist.
  2016-09-23 14:51:33 naarad.utils [WARNING] /tmp/samza-tests-20160923144307/scripts/naarad.cfg : file does not exist.
  2016-09-23 14:51:33 smoke_tests [INFO] Running validate_samza_job
  2016-09-23 14:51:33 performance_tests [INFO] Running validate_kafka_read_write_performance
  2016-09-23 14:52:13 zopkio.remote_host_helper [INFO] stopping resourcemanager
  2016-09-23 14:52:14 zopkio.remote_host_helper [INFO] Stopping zookeeper ... STOPPED
  2016-09-23 14:52:20 zopkio.remote_host_helper [INFO] stopping nodemanager
  nodemanager did not stop gracefully after 5 seconds: killing with kill -9
  2016-09-23 14:52:21 zopkio.test_runner [INFO] Execution of configuration: single execution complete
  2016-09-23 14:52:21 zopkio.test_runner [INFO] test_samza_job----passed
  2016-09-23 14:52:21 zopkio.test_runner [INFO] test_container_performance----passed
  2016-09-23 14:52:21 zopkio.test_runner [INFO] test_kafka_read_write_performance----passed
  2016-09-23 14:52:21 zopkio [INFO] Exiting zopkio
  ```
- [x] deploy a job in staging on a Kafka 0.8.x cluster
- [ ] deploy a job in production on a Kafka 0.8.x cluster
- [ ] deploy a job in staging on a Kafka 0.10.x cluster
- [ ] deploy a job in production on a Kafka 0.10.x cluster

### CODE REVIEW

File: samza-test/src/main/python/configs/tests.json [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {
-  "samza_executable": "samza-test_2.10-0.11.0-SNAPSHOT.tgz",
+  "samza_executable": "samza-test_2.10-0.10.1-SNAPSHOT.tgz",
```
Comment (nickpan47) [inline samza-test/src/main/python/configs/tests.json]: The latest samza release will go w/ samza-test_2.10-0.12.0-SNAPSHOT.tgz soon. Please check master and update accordingly.

File: samza-test/src/test/scala/org/apache/samza/test/integration/StreamTaskTestUtil.scala [hunk 2]
```diff
@@ -122,12 +104,36 @@ object StreamTaskTestUtil {
...
   def beforeSetupServers {
-    zookeeper = new EmbeddedZookeeper(zkConnect)
-    server1 = TestUtils.createServer(new KafkaConfig(props1))
-    server2 = TestUtils.createServer(new KafkaConfig(props2))
-    server3 = TestUtils.createServer(new KafkaConfig(props3))
-    zkClient = new ZkClient(zkConnect + "/", 6000, 6000, ZKStringSerializer)
+    zookeeper = new EmbeddedZookeeper()
+    zkUtils = ZkUtils(zkConnect, zkSessionTimeout, zkConnectionTimeout, JaasUtils.isZkSecurityEnabled())
+
+    val props = TestUtils.createBrokerConfigs(3, zkConnect, true)
```
Comment (nickpan47) [inline samza-test/src/test/scala/org/apache/samza/test/integration/StreamTaskTestUtil.scala]: One thing that I am curious about the broker config, did you turn on the new message format w/ timestamp support at the broker side in this patch?

Comment (ncolomer) [inline samza-test/src/test/scala/org/apache/samza/test/integration/StreamTaskTestUtil.scala]: Kafka distrib downloaded by zopkio is [kafka_2.10-0.10.0.1.tgz](http://archive.apache.org/dist/kafka/0.10.0.1/kafka_2.10-0.10.0.1.tgz) from Apache repositories.
The 0.10.x protocol for both internal and client communication [is enabled by default](https://kafka.apache.org/documentation.html#upgrade_10) for this version. Kafka lets you lower it for compatibility/upgrade purpose (our case) but the Kafka 10 client is fortunately backward compatible.

File: samza-test/src/main/python/configs/tests.json (hunk 1 above)
Comment (nickpan47) [inline samza-test/src/main/python/configs/tests.json]: Makes sense. Let me sync w/ Xinyu Liu on the latest check-ins to master regarding to the version generated from trunk.

Meanwhile, I would encourage you to rebase again the latest master and validate that the version here still works. Thanks!

File: samza-test/src/test/scala/org/apache/samza/test/integration/StreamTaskTestUtil.scala (hunk 2 above)
Comment (nickpan47) [inline samza-test/src/test/scala/org/apache/samza/test/integration/StreamTaskTestUtil.scala]: Thanks! Sounds great!

File: samza-test/src/main/python/configs/tests.json (hunk 1 above)
Comment (ncolomer) [inline samza-test/src/main/python/configs/tests.json]: Yup, I do a rebase on upstream/master every time I push on this branch.

**Comment by ncolomer** [comment]
We finally decided to not upgrade Samza to current versions (either 0.10, 0.11 or 0.12 master) but to stay on 0.9.1 instead since we rely on a legacy YARN cluster (cdh5.4.0 distribution) and there were (sooo) many conflicts and issues to deal with (we tried from surgical maven exclusions to dependency shading).

What we did is we checked out Samza 0.9.1 branch, back ported this PR's changes and built a custom 0.9.2-rc1 version that we host on our private repo. We made all our jobs reach both staging and production using this patched version. All run smoothly to date.

You can take a look at the 0.9.1 patch [here](https://github.com/ncolomer/samza/compare/0.9.1...ncolomer:samza91-kafka10), it might be useful for those in the same situation.

About this PR, I'm afraid we have no way to experience it on our production due to the reasons mentioned above. We planned to migrate our Yarn cluster too (probably on a Mesos runtime) but it should not happen in a near future.

File: gradle/dependency-versions.gradle [hunk 3]
```diff
@@ -25,10 +25,10 @@
   junitVersion = "4.8.1"
   mockitoVersion = "1.8.4"
   scalaTestVersion = "2.2.4"
-  zkClientVersion = "0.3"
+  zkClientVersion = "0.9"
```
Comment (shanthoosh) [inline gradle/dependency-versions.gradle]: Just curious. Kafka 0.10.0.1 version depends upon zkClient version 0.8. https://mvnrepository.com/artifact/org.apache.kafka/kafka_2.11/0.10.0.1. Any reason for using zkClient with version 0.9 ?

File: docs/learn/tutorials/versioned/samza-rest-getting-started.md [hunk 4]
```diff
@@ -48,7 +48,7 @@ Run the following commands:
 {% highlight bash %}
 cd samza-rest/build/distributions/
 mkdir -p deploy/samza-rest
-tar -xvf ./samza-rest-0.10.1-SNAPSHOT.tgz -C deploy/samza-rest
+tar -xvf ./samza-rest-0.11.0-SNAPSHOT.tgz -C deploy/samza-rest
```
Comment (shanthoosh) [inline docs/learn/tutorials/versioned/samza-rest-getting-started.md]: There is a merge conflict here. In current master this line is of this form: 
`tar -xvf ./samza-rest-0.11.1-SNAPSHOT.tgz -C deploy/samza-rest`. 
You can rebase with recent master to fix this.

Comment (shanthoosh) [inline docs/learn/tutorials/versioned/samza-rest-getting-started.md]: We've tested and validated this kafka patch with internal samza build at LinkedIn. This looks good.

**Comment by nickpan47** [comment]
@ncolomer we have merged this one into 0.12.0. This PR can be closed. Thanks!

**Comment by ncolomer** [comment]
@nickpan47 that's great news! thanks

**Comment by aries-zhang** [comment]
@nickpan47 Do you have any time table on releasing 0.12.0? We are trying to use Samza on Aliyun, which is a Chinese large cloud provider and their message service recently started to support standard kafka 0.10 protocol as an internal release. If we can directly upgrade to Samza 0.12.0 before their formal release, which I suppose will happen in two or three months, it will save our team a lot of time and work and will be worth waiting.

**Comment by nickpan47** [comment]
Hi, Aries,

That's awesome news! The expected timeline for Samza 0.12 should be within
a month. I think it will well fit in your timeline!

Cheers!

-Yi

**Comment by aries-zhang** [comment]
@nickpan47 Really good news, cheers!

**Comment by vjagadish** [comment]
Closing this PR (as Samza 0.12 contains this patch). Thank you @ncolomer for the patch!

**Comment by nickpan47** [comment]
@vjagadish1989 did you close this one? This PR can be closed since it was already merged in 0.12. @ncolomer could you close this PR? Thanks!

**Comment by vjagadish1989** [comment]
Unfortunately, I could not find ways to close this from the github UI. (Maybe, I don't have accesses on the github repo provisioned correctly). Also, I believe the mirror is read-only. 

From some misc docs [here](https://cwiki.apache.org/confluence/display/REEF/Accepting+Pull+Requests): 
Further, the mirror on GitHub is read-only. Which means we can neither accept nor close pull requests filled there via the GitHub web UI.

**Comment by ncolomer** [comment]
Seems I can close it, so closing it!
Thanks @nickpan47 @vjagadish @vjagadish1989

**Pull request closed without merging**